File: src/main/java/cspfj/constraint/extension/MatrixGeneral.java
package cspfj.constraint.extension;

import java.util.Arrays;

public final class MatrixGeneral implements Matrix, Cloneable {
    private int[] skip;

    private boolean[] matrix;

    private boolean empty;

    public MatrixGeneral(final int[] sizes, final boolean initialState) {
        int nbValues = 1;
        for (int size : sizes) {
            nbValues *= size;
        }

        matrix = new boolean[nbValues];
        Arrays.fill(matrix, initialState);

        skip = new int[sizes.length - 1];
        skip[0] = sizes[0];
        for (int i = 1; i < skip.length; i++) {
            skip[i] = skip[i - 1] * sizes[i];
        }
        empty = !initialState;
        // System.out.println(Arrays.toString(skip));
    }

    @Override
    public boolean check(final int[] tuple) {
        return matrix[matrixIndex(tuple)];
    }

    @Override
    public void set(final int[] tuple, final boolean status) {
        matrix[matrixIndex(tuple)] = status;
        empty = false;
    }

    private int matrixIndex(final int[] tuple) {
        int index = tuple[0];
        for (int i = skip.length; --i >= 0;) {
            index += skip[i] * tuple[i + 1];
        }
        return index;
    }

    @Override
    public MatrixGeneral clone() throws CloneNotSupportedException {
        final MatrixGeneral clone = (MatrixGeneral) super.clone();
        clone.skip = skip.clone();
        clone.matrix = this.matrix.clone();
        return clone;

    }

    @Override
    public boolean isEmpty() {
        return empty;
    }

}

File: src/main/scala/cspfj/constraint/extension/MatrixGeneral.scala
package cspfj.constraint.extension;

import java.util.Arrays;

final class MatrixGeneral(sizes: Seq[Int], val initialState: Boolean) extends Matrix with Cloneable {
  require(sizes.map(BigInt(_)).product <= Int.MaxValue)
  
  var matrix = Array.fill(sizes.product)(initialState)

  var skip = sizes.scan(1)(_ * _).init.toArray

  var empty = !initialState

  def check(tuple: Array[Int]) = matrix(matrixIndex(tuple))

  def set(tuple: Array[Int], status: Boolean) {
    matrix(matrixIndex(tuple)) = status
    empty = false
  }

  private def matrixIndex(tuple: Array[Int]) = {
    //(tuple, skip).zipped.map((t, s) => t * s).sum
    var s = 0;
    var i = tuple.length - 1;
    while (i >= 0) {
      s += tuple(i) * skip(i)
      i -= 1
    }
    s
  }

  def copy = {
    val clone = super.clone.asInstanceOf[MatrixGeneral]
    clone.matrix = matrix.clone
    clone
  }

  def isEmpty = empty
}
